test/org/apache/tomcat/websocket/TestWsSessionSuspendResume.java

```diff
@@ -0,0 +1,144 @@
+        Tomcat tomcat = getTomcatInstance();
+
+        Context ctx = tomcat.addContext("", null);
+        ctx.addApplicationListener(Config.class.getName());
+
+        Tomcat.addServlet(ctx, "default", new DefaultServlet());
+        ctx.addServletMappingDecoded("/", "default");
+
+        tomcat.start();
+
+        WebSocketContainer wsContainer = ContainerProvider.getWebSocketContainer();
+
+        ClientEndpointConfig clientEndpointConfig = ClientEndpointConfig.Builder.create().build();
+        Session wsSession = wsContainer.connectToServer(
+                TesterProgrammaticEndpoint.class,
+                clientEndpointConfig,
+                new URI("ws://localhost:" + getPort() + Config.PATH));
+
+        CountDownLatch latch = new CountDownLatch(2);
+        wsSession.addMessageHandler(String.class, message -> {
+            Assert.assertTrue("[echo, echo, echo]".equals(message));
+            latch.countDown();
+        });
+        for (int i = 0; i < 8; i++) {
+            wsSession.getAsyncRemote().sendText("echo");
+        }
+
+        boolean latchResult = latch.await(30, TimeUnit.SECONDS);
+        Assert.assertTrue(latchResult);
+
+        wsSession.close();
+    }
+
+
+    public static final class Config extends TesterEndpointConfig {
+        private static final String PATH = "/echo";
+
+        @Override
+        protected Class<?> getEndpointClass() {
+            return SuspendResumeEndpoint.class;
+        }
+
+        @Override
+        protected ServerEndpointConfig getServerEndpointConfig() {
+            return ServerEndpointConfig.Builder.create(getEndpointClass(), PATH).build();
+        }
+    }
+
+
+    public static final class SuspendResumeEndpoint extends Endpoint {
+
+        @Override
+        public void onOpen(Session session, EndpointConfig  epc) {
+            MessageProcessor processor = new MessageProcessor(session, 3);
+            session.addMessageHandler(String.class, message -> processor.addMessage(message));
+        }
+
+        @Override
+        public void onClose(Session session, CloseReason closeReason) {
+            try {
+                session.close();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+
+        @Override
+        public void onError(Session session, Throwable t) {
+            t.printStackTrace();
+        }
+    }
+
+
+    private static final class MessageProcessor {
+        private final Session session;
+        private final int count;
+        private final List<String> messages = new ArrayList<>();
+
+        MessageProcessor(Session session, int count) {
+            this.session = session;
+            this.count = count;
+        }
+
+        void addMessage(String message) {
+            if (messages.size() == count) {
+                ((SuspendableMessageReceiver) session).suspend();
+                session.getAsyncRemote().sendText(messages.toString(), result -> {
```

Comment:
IMO `sendText(...)` should be executed in a different thread with some delay.
Otherwise now even without calling `suspend()` the behavior would be same. `messages` will be serialized to its String form and the fourth message would have a chance to break the test.